go/types/opcache_compare.go

```diff
@@ -6,39 +6,124 @@ package types
 
 import (
 	"bytes"
+	"encoding/binary"
 
 	"github.com/attic-labs/noms/go/d"
 	"github.com/attic-labs/noms/go/hash"
 )
 
 type opCacheComparer struct{}
 
-func (opCacheComparer) Compare(a, b []byte) int {
+func (opCacheComparer) Compare(a, b []byte) (res int) {
 	if res := bytes.Compare(a[:uint32Size], b[:uint32Size]); res != 0 {
 		return res
 	}
-	a, b = a[uint32Size:], b[uint32Size:]
+	return compareEncodedKeys(a[uint32Size:], b[uint32Size:])
+}
+
+func (opCacheComparer) Name() string {
+	return "noms.OpCacheComparator"
+}
+
+func (opCacheComparer) Successor(dst, b []byte) []byte {
+	return nil
+}
+
+func (opCacheComparer) Separator(dst, a, b []byte) []byte {
+	return nil
+}
 
+func compareEncodedKeys(a, b []byte) int {
 	if compared, res := compareEmpties(a, b); compared {
 		return res
 	}
-	aKind, bKind := NomsKind(a[0]), NomsKind(b[0])
-	switch aKind {
-	default:
-		if bKind <= StringKind {
+
+	// keys are encoded as either values:
+	//   nomsKind(1-byte) + serialized len(4-bytes) + serialized value(n-bytes)
+	// or digests:
+	//   nomsKind(1-byte) + digest(hash.Bytelen-bytes)
+	splitAfterFirstKey := func(bs []byte) ([]byte, []byte) {
+		keyLen := 1 + hash.ByteLen
+		if isKindOrderedByValue(NomsKind(bs[0])) {
+			l := int(binary.BigEndian.Uint32(bs[1:5]))
+			keyLen = 1 + uint32Size + l
+		}
+		return bs[:keyLen], bs[keyLen:]
+	}
+
+	// a[0] and b[0] represent NomsKind of leafNode being operated on
+	// a[1] and b[1] are the number of keys encoded in this byte slice
+	numAGraphKeys, numBGraphKeys := a[1], b[1]
+	minNumKeys := minByte(numAGraphKeys, numBGraphKeys)
+
+	a, b = a[2:], b[2:]
+	cres := 0
+	for pos := 0; pos < int(minNumKeys) && cres == 0; pos++ {
+		aKey, aRest := splitAfterFirstKey(a)
+		bKey, bRest := splitAfterFirstKey(b)
+		cres = compareEncodedKey(aKey, bKey)
+		a, b = aRest, bRest
+	}
+
+	if cres == 0 {
+		if numAGraphKeys < numBGraphKeys {
+			return -1
+		}
+		if numAGraphKeys > numBGraphKeys {
 			return 1
 		}
-		a, b = a[1:], b[1:]
+	}
+	return cres
+}
+
+// compareKey accepts two bytestreams that each contain a number of encoded keys.
+// It compares the first key in the encoding and returns the comparison result for
+// that key along with two slices that hold the bytes following the current key
+func compareEncodedKey(a, b []byte) (res int) {
+	// each key is encoded as NomsKind(1-byte) + length(4-bytes) + encoding(n-bytes)
+	aKind, bKind := NomsKind(a[0]), NomsKind(b[0])
+	if isKindOrderedByValue(aKind) || isKindOrderedByValue(bKind) {
+		if res := compareKinds(aKind, bKind); res != 0 {
+			return res
+		}
+	}
+
+	// we know the noms kinds are identical. If they are both hashes just compare
+	// the hashes
+	if !isKindOrderedByValue(aKind) && !isKindOrderedByValue(bKind) {
```

Comment:
I don't think I can easily switch positions here. The second "simpler" case relies on knowing that the noms kinds are identical.